File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
import sys
import copy
for test_case in range(1):
    n, m = map(int, sys.stdin.readline().split())

    string = list(sys.stdin.readline().split())
    ans = []
    tmp = []
    visited = [False for i in range(m)]
    mo_list = ['a', 'e', 'i', 'o', 'u']
    def backtracking():
        if len(tmp) == n:
            ja, mo = 0, 0

            for alpha in tmp:
                if alpha in mo_list:
                    ja += 1
                else:
                    mo += 1

            if ja >= 1 and mo >= 2:
                ans.append(copy.deepcopy(tmp))

            return

        for i in range(m):
            if not tmp:
                tmp.append(string[i])
                visited[i] = True
                backtracking()
                visited[i] = False
                tmp.pop()
                continue

            if ord(tmp[-1]) < ord(string[i]) and not visited[i]:
                tmp.append(string[i])
                visited[i] = True
                backtracking()
                visited[i] = False
                tmp.pop()

    backtracking()
    ans.sort()
    for a in ans:
        print(''.join(a))

File: 백준/Gold/1759. 암호 만들기/암호 만들기.java
import java.util.*;
import java.io.*;

public class Main
{
	static String[] vowel = {"a", "e", "i", "o", "u"};
	static ArrayList<String> tmp = new ArrayList<String>();
	static ArrayList<String> cases = new ArrayList<String>();
	static String[] alpha = null;
	static int l = 0;
	static int c = 0;
	public static void main(String args[]) throws Exception
	{
//		BufferedReader br = new BufferedReader(new FileReader("./input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));	
		
		String[] str = br.readLine().split(" ");
		l = Integer.parseInt(str[0]);
		c = Integer.parseInt(str[1]);
		
		alpha = br.readLine().split(" ");
		Arrays.sort(alpha);
		backtracking(0);
		for ( String cas: cases ) {
			System.out.println(cas);
		}
	}
	
	static void backtracking(int p) {
		if (tmp.size() == l) {
			int ja = 0;
			int mo = 0;
			for (String alphabet : tmp) {
				if (Arrays.asList(vowel).contains(alphabet)) mo += 1;
				else ja += 1;
			}
			if ( mo >= 1 && ja >= 2) cases.add(String.join("", tmp));
			return;
		}
		
		for (int i = p; i < c; i++) {
			if (tmp.size() == 0) {
				tmp.add(alpha[i]);
				backtracking(i+1);
				tmp.remove(tmp.size() - 1);
				continue;
			} 

			tmp.add(alpha[i]);
			backtracking(i+1);
			tmp.remove(tmp.size() - 1);			
		}
	}
}